lib/logdna/client.rb

```diff
@@ -44,95 +63,134 @@ def process_message(msg, opts = {})
+        end
+        @buffer.push(Message.new(processed_message, total_size))
+
+        if total_size >= @flush_threshold
+          unschedule_flush
           flush
         else
           schedule_flush
         end
-      else
-        @side_message_lock.synchronize do
-          @side_messages.push(process_message(msg, opts))
-        end
       end
     end
 
-    # This method has to be called with @lock
-    def send_request
-      @side_message_lock.synchronize do
-        @buffer.concat(@side_messages)
-        @side_messages.clear
+    ##
+    # Flushes all logs to LogDNA asynchronously
+    def flush(options = {})
+      Concurrent::Future.execute({ executor: @work_thread_pool }) { flush_sync(options) }
+    end
+
+    ##
+    # Flushes all logs to LogDNA synchronously
+    def flush_sync(options = {})
+      slices = @lock.synchronize do
+        # Slice the buffer into chunks that try to be no larger than @flush_limit. Slice points are found with
+        # a binary search thanks to the structure of @buffer. We are working backwards because it's cheaper to 
+        # remove from the tail of an array instead of the head
+        slices = []
+        while @buffer.length != 0 do
+          search_size = @buffer[-1].size - @flush_limit
+          if search_size < 0
+            search_size = 0
+          end
+
+          slice_index = @buffer.bsearch_index { |message| message.size >= search_size }
+          slices.push(@buffer.pop(@buffer.length - slice_index).map { |message| message.source })
+        end
+        slices
+      end
+
+      # Remember the chunks are in reverse order, this un-reverses them
+      slices.reverse_each do |slice|
+        if options[:block_on_requests]
+          send_request(slice)
+        else
+          Concurrent::Future.execute({ executor: @request_thread_pool }) { send_request(slice) }
+        end
       end
+    end
 
+    def send_request(slice)
+      # TODO: Remove instance-global request object
       @request.body = {
         e: "ls",
-        ls: @buffer
+        ls: slice
       }.to_json
 
-      handle_exception = lambda do |message|
-        @internal_logger.debug(message)
-        @exception_flag = true
-        @side_message_lock.synchronize do
-          @side_messages.concat(@buffer)
-        end
-      end
+      flush_id = "#{Random.uuid} [#{slice.length} lines]"
+      error_header = "Flush {#{flush_id}} failed."
+      tries = 0
+      loop do
+        tries += 1
 
-      begin
-        @response = Net::HTTP.start(
-          @uri.hostname,
-          @uri.port,
-          use_ssl: @uri.scheme == "https"
-        ) do |http|
-          http.request(@request)
-        end
-        if @response.is_a?(Net::HTTPForbidden)
-          @internal_logger.debug("Please provide a valid ingestion key")
-        elsif !@response.is_a?(Net::HTTPSuccess)
-          handle_exception.call("The response is not successful #{@response}")
+        if tries >= 4
```

Comment:
would avoid magic numbers, this can likely go in `Resources` and perhaps make it an option

Reply:
Moving this out to `Resources`.